### PR TITLE
[FLOC-1937] Change testing instructions to remove need for shared staging branch

### PR DESCRIPTION
**Pull request by jongiddy**
https://clusterhq.atlassian.net/browse/FLOC-1937

Procedure to start a Buildbot master without using the `staging` branch.

[Review on Reviewable](https://reviewable.io/reviews/clusterhq/build.clusterhq.com/92)

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -82,6 +64,21 @@ This command will display the external IP address of the EC2 instance.
...
 Run ``python start-aws.py --help`` to see the available options to this command.
 
+Create a staging Docker image
+=============================
+
+On the staging server, run the following commands::
+
+   [aws]$ sudo yum install -y docker-io fabric git
+   [aws]$ sudo setenforce 0
+   [aws]$ sudo systemctl start docker
```
Comment (wallrj) [inline README.rst]: The commands above belong in the "Create a staging server" section I think.

File: README.rst [hunk 2]
```diff
@@ -82,6 +64,21 @@ This command will display the external IP address of the EC2 instance.
 
 Run ``python start-aws.py --help`` to see the available options to this command.
 
+Create a staging Docker image
+=============================
+
+On the staging server, run the following commands::
```
Comment (wallrj) [inline README.rst]: :question: Maybe `.. code-block:: console` although GitHub didn't seem to do the highlighting when I tried it for the new Openstack build slave instructions.

File: README.rst [hunk 3]
```diff
@@ -82,6 +64,21 @@ This command will display the external IP address of the EC2 instance.
...
+
+   [aws]$ sudo yum install -y docker-io fabric git
+   [aws]$ sudo setenforce 0
+   [aws]$ sudo systemctl start docker
+   [aws]$ git clone https://github.com/ClusterHQ/build.clusterhq.com.git
+   [aws]$ cd build.clusterhq.com
+   [aws]$ # Change <BRANCH> to the branch of build.clusterhq.com you want
+   [aws]$ git checkout <BRANCH>
+   [aws]$ sudo docker build -t clusterhq/build.clusterhq.com:staging .
```
Comment (wallrj) [inline README.rst]: :question: I'd prefer `--tag` for clarity.

Comment (wallrj) [inline README.rst]: Also add the `--rm` option to remove the intermediate  images.

File: README.rst [hunk 4]
```diff
@@ -82,6 +64,21 @@ This command will display the external IP address of the EC2 instance.
...
+   [aws]$ sudo docker run --name buildmaster-data -v /srv/buildmaster/data busybox /bin/true
```
Comment (wallrj) [inline README.rst]: I never did this step. How important is it? If it's important, add an explanatory note. I guess it wipes out the data from the existing staged build slave (if there is one on this staging server)

File: README.rst [hunk 5]
```diff
@@ -82,6 +64,21 @@ This command will display the external IP address of the EC2 instance.
...
+
```
Comment (wallrj) [inline README.rst]: :question: I found `fab:getConfig` useful to ensure that I'm testing the latest config from lastpass. Maybe mention that below. (unless it's mentioned elsewhere.)

File: README.rst [hunk 6]
```diff
@@ -99,21 +96,24 @@ Make the following changes to the ``staging.yml`` file:
...
-   $ fab update:staging.yml
-
-Log in to the EC2 instance with the credentials from the ``auth`` section of the config file.
+Connect to the IP address of the EC2 instance and log in to the Buildmaster portal with the credentials from the ``auth`` section of the config file.
+Click on the ``flocker`` link to access the web form.
 
 The staging setup is missing the ability to trigger builds in response to Github pushes.
+To trigger a build, enter a branch name and click the ``Force`` button to start testing a Flocker branch.
+
+The staging master will start latent slaves on AWS EC2 automatically when builds have been triggered.
```
Comment (wallrj) [inline README.rst]: :question: I always wondered about this. What will the names of the latent slaves be? Don't they end up clashing with the ones built by the production master? I think you can probably have duplicate server names on AWS and Rackspace...but perhaps we should create an issue to more easily identify the master that a latent slave is controlled by, from its name.

File: README.rst [hunk 7]
```diff
@@ -99,21 +96,24 @@ Make the following changes to the ``staging.yml`` file:
 Start staging server
 ====================
 
-To start a Buildbot master on this machine run::
-
-   $ fab start:staging.yml
+Once the Docker image has built on the staging server, and the staging.yml file has been created, start the test Buildbot master from the local machine using::
 
-To update a slave on this machine, run::
+   $ fab restart:staging.yml
```
Comment (wallrj) [inline README.rst]: Using `restart` is a bit of a hack to avoid pulling an image from dockerhub. Maybe explain that here.

**Comment by wallrj** [comment]
Thanks @jongiddy 

This looks pretty good. 

 * It matches the process that I've been using recently https://github.com/ClusterHQ/build.clusterhq.com/pull/76#issuecomment-104010081
 * I think it could do with one or two explanatory notes (see inline comments above)
 * There are merge conflicts. Branch needs merging forward.
 * Since the `staging`  branch instructions are now removed, should we remove
   * the automatic builder from Docker Hub
   * the staging.build.clusterhq.com server
   * any build slaves that go with that server.
 * @tomprince shared a git post-receive script which allows him to automate the docker image rebuilding for each git push.

```
#!/usr/bin/bash

project_name=$(basename $(pwd))
tag="$project_name:$(git log --pretty=format:'%h' -n 1)"
checkout=$(mktemp -d)
git archive HEAD | tar -x -C $checkout
find $checkout -exec touch -d 2014-01-01 '{}' ';'
echo "--> Building $project_name to $tag"
docker build -t clusterhq/build.clusterhq.com:staging $checkout
rm -rf $checkout
if [ $? -ne 0 ]; then
  echo '--! Build failed, aborting'
fi
```

https://zulip.com/#narrow/near/39854315/stream/engineering/topic/request.20to.20be.20added.20to.20clusterhq.20dockerhub

Discuss this with @tomprince then either:
 * Create follow up issues for further improvements and put this up for a quick final review,
 * or update the instructions to incorporate Tom's post-receive script and and I'll test that.

File: README.rst [hunk 8]
```diff
@@ -99,21 +96,24 @@ Make the following changes to the ``staging.yml`` file:
...
+Click on the ``flocker`` link to access the web form.
 
 The staging setup is missing the ability to trigger builds in response to Github pushes.
+To trigger a build, enter a branch name and click the ``Force`` button to start testing a Flocker branch.
+
+The staging master will start latent slaves on AWS EC2 automatically when builds have been triggered.
 
-The staging master will start Linux slaves on AWS EC2 automatically.
-To start a Mac OS X slave, see below.
+Tests that require Mac OS X or to start VM's cannot use AWS EC2 latent slaves.
```
Comment (adamtheturtle) [inline README.rst]: "Tests that require to start VMs" or "Tests that to start VMs" doesn't make sense.

File: README.rst [hunk 9]
```diff
@@ -82,6 +64,21 @@ This command will display the external IP address of the EC2 instance.
 
 Run ``python start-aws.py --help`` to see the available options to this command.
 
+Create a staging Docker image
+=============================
```
Comment (adamtheturtle) [inline README.rst]: Maybe some of this can be part of start-aws?

Separate what has to be done only once and what is done each time.

File: README.rst (hunk 3 above)
Comment (adamtheturtle) [inline README.rst]: What did I do wrong @wallrj ?

```
[fedora@ip-172-31-40-6 build.clusterhq.com]$ sudo docker build --tag --rm clusterhq/build.clusterhq.com:staging .
docker: "build" requires 1 argument. See 'docker build --help'.
```

File: README.rst (hunk 5 above)
Comment (adamtheturtle) [inline README.rst]: Already mentioned in "Create the configuration"

**Comment by adamtheturtle** [comment]
This isn't perfect but I've fixed the worst comments and people are using these instructions (they're a big improvement on the current system) so I think that this should be merged.

File: README.rst (hunk 6 above)
Comment (jml) [inline README.rst]: "latent slave" is a weird terminology (I don't really know what it means). However, it's used widely throughout this doc, so should be fixed in a different patch.

**Comment by jml** [comment]
:+1: I agree. Merge when ready.

**Comment by adamtheturtle** [comment]
> [we should remove] staging.build.clusterhq.com

I think that this is just one of the multiple staging build servers we have (and maybe the one Tom uses).

As for removing the Docker Hub builders, I don't think it is necessary.

File: README.rst (hunk 6 above)
Comment (jongiddy) [inline README.rst]: The slave names do not indicate the master, but each tag has a Buildmaster tag, which does. This can be used to restrict the view just to the appropriate slaves.

Comment (jongiddy) [inline README.rst]: "Latent slave" is a slave that gets started when tests are waiting to run, and gets stopped when no further tests are waiting to run (on that platform).  As opposed to a small number of dedicated slaves (one Linux box for running the Vagrant tests and one Mac).
